unison-cli/src/Unison/Share/Sync.hs

```diff
@@ -606,64 +619,141 @@ getCausalHashByPath httpClient unisonShareUrl repoPath =
 uploadEntities ::
   AuthenticatedHttpClient ->
   BaseUrl ->
-  Sqlite.Connection ->
+  (forall a. (Sqlite.Connection -> IO a) -> IO a) ->
   Share.RepoName ->
   NESet Hash32 ->
-  (Int -> Int -> IO ()) ->
+  UploadProgressCallbacks ->
   IO Bool
-uploadEntities httpClient unisonShareUrl conn repoName hashes0 uploadProgressCallback =
-  loop 0 hashes0
+uploadEntities httpClient unisonShareUrl connect repoName hashes0 callbacks = do
+  hashesVar <- newTVarIO (NESet.toSet hashes0)
+  -- FIXME document this
+  dedupeVar <- newTVarIO Set.empty
+  nextWorkerIdVar <- newTVarIO 0
+  workersVar <- newTVarIO Set.empty
+  workerFailedVar <- newEmptyTMVarIO
```

Comment:
Hrmm, `ki` propagates exceptions from threads to the parent right?
I wonder if we should just throw an exception on failure rather than need to remember to manually check a failure var every time before we do something.

I notice lower down you only bail on failure if there isn't more work to do, is there a particular reason for that?

Exiting with an exception would maybe mean we have some entities downloaded that we didn't insert, but honestly that seems like a reasonable sacrifice to make for simplicity and avoiding a foot-gun of forgetting to check for errors and rolling forward in some sort of invalid state.

Reply:
Hmm... I'm not sure I agree. I think returning failures from the endpoint as 200s so we could pattern match on them was done so the code is easier to understand, and we don't have to know what exception types might be thrown, and where.

I agree it's confusing right now -- maybe we shouldn't have a `catchSyncExceptions` thing that only catches some subset of  exceptions these actions might throw? Couldn't we return these as `Left`s too, and pass them around manually? (Or use `ExceptT` syntactic sugar).